pubsub/streaming-analytics/PubSubToGCS_test.py

```diff
@@ -12,89 +12,68 @@
-def publisher_client():
-    yield pubsub_v1.PublisherClient()
-
-
-@pytest.fixture
-def topic_path(publisher_client):
-    topic_path = publisher_client.topic_path(PROJECT, TOPIC)
-
-    try:
-        publisher_client.delete_topic(topic_path)
-    except Exception:
-        pass
-
-    response = publisher_client.create_topic(topic_path)
-    yield response.name
-
-
-def _infinite_publish_job(publisher_client, topic_path):
-    while True:
-        future = publisher_client.publish(
-            topic_path, data='Hello World!'.encode('utf-8'))
-        future.result()
-        time.sleep(10)
-
-
-def test_run(publisher_client, topic_path):
-    """This is an integration test that runs `PubSubToGCS.py` in its entirety.
-    It checks for output files on GCS.
-    """
-
-    # Use one process to publish messages to a topic.
-    publish_process = mp.Process(
-        target=lambda: _infinite_publish_job(publisher_client, topic_path))
-
-    # Use another process to run the streaming pipeline that should write one
-    # file to GCS every minute (according to the default window size).
-    pipeline_process = mp.Process(
-        target=lambda: sp.call([
-            'python', 'PubSubToGCS.py',
-            '--project', PROJECT,
-            '--runner', 'DirectRunner',
-            '--temp_location', tempfile.mkdtemp(),
-            '--input_topic', topic_path,
-            '--output_path', 'gs://{}/pubsub/{}/output'.format(BUCKET, UUID),
-        ])
-    )
-
-    publish_process.start()
-    pipeline_process.start()
-
-    # Times out the streaming pipeline after 90 seconds.
-    pipeline_process.join(timeout=90)
-    # Immediately kills the publish process after the pipeline shuts down.
-    publish_process.join(timeout=0)
-
-    pipeline_process.terminate()
-    publish_process.terminate()
-
-    # Check for output files on GCS.
-    gcs_client = beam.io.gcp.gcsio.GcsIO()
-    # This returns a dictionary.
-    files = gcs_client.list_prefix('gs://{}/pubsub/{}'.format(BUCKET, UUID))
-    assert len(files) > 0
-
-    # Clean up. Delete topic. Delete files.
-    publisher_client.delete_topic(topic_path)
-    gcs_client.delete_batch(list(files))
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.testing.test_pipeline import TestPipeline
+from apache_beam.testing.test_stream import TestStream
+from apache_beam.testing.test_utils import TempDir
+from apache_beam.transforms.window import TimestampedValue
+
+import PubSubToGCS
+
+PROJECT = os.environ["GCLOUD_PROJECT"]
+BUCKET = os.environ["CLOUD_STORAGE_BUCKET"]
+UUID = uuid.uuid1().hex
+
+
+class PubSubToGCSTest(unittest.TestCase):
+    def test_pubsub_to_gcs(self):
+
+        pipeline_options = PipelineOptions(
+            project=PROJECT,
+            runner="DirectRunner",
+            temp_location=TempDir().get_path(),
+            streaming=True,
+            save_main_session=True,
+        )
+
+        with TestPipeline(options=pipeline_options) as p:
```

Comment:
Ideally, we want to use the same function as in the code sample, so if it ever changes the test is a safeguard to know if it breaks or not. You can `mock` the `beam.Pipeline` symbol to use `TestPipeline` instead. Then you can also mock the `beam.io.ReadFromPubSub` symbol to use `TestStream` instead.

Then you can call your `PubSubToGCS.run` function with the rest of the options.

Here's an example where I mocked `Pipeline` and other functions.
https://github.com/apache/beam/blob/master/sdks/python/apache_beam/examples/snippets/transforms/elementwise/pardo_test.py

I think something like this should do it for you:

```py
@mock.patch('apache_beam.Pipeline', TestPipeline)
@mock.patch('apache_beam.io.ReadFromPubSub', lambda _: (
    TestStream()
    .advance_watermark_to(0)
    ...
))
def test_pubsub_to_gcs(self):
    PubSubToGCS.run(
        input_topic='unused',  # mocked by TestStream
        output_path="gs://{}/pubsub/{}/output".format(BUCKET, UUID),
        window_size=1,  # minutes
        pipeline_args=[
            '--project', PROJECT,
            '--temp_location', TempDir().get_path(),
        ],
    )

    # Check for output files on GCS.
    ...

    # Clean up.
    ...
```

There is no need to specify the `DirectRunner` explicitly since it's always the default option if no runner is selected :)

Reply:
`_` didn't work but specifically calling out `topic` in the lambda function works. 😃